src/AcceptanceTests/SagaDataStorage/When_storing_saga_with_high_contention.cs

```diff
@@ -0,0 +1,165 @@
+
+            public Stopwatch Watch { get; } = new Stopwatch();
+
+            public TimeSpan Elapsed => Watch.Elapsed;
+
+            public int NumberOfMessages { get; } = 20;
+
+            public long NumberOfRetries => Interlocked.Read(ref numberOfRetries);
+
+            public void IncrementNumberOfRetries()
+            {
+                Interlocked.Increment(ref numberOfRetries);
+            }
+        }
+
+        class SagaEndpoint : EndpointConfigurationBuilder
+        {
+            public SagaEndpoint()
+            {
+                EndpointSetup<DefaultServer, Context>((b, c) =>
+                {
+                    b.LimitMessageProcessingConcurrencyTo(c.NumberOfMessages);
+                    var recoverability = b.Recoverability();
+                    recoverability.Immediate(s =>
+                    {
+                        s.OnMessageBeingRetried(m =>
+                        {
+                            c.IncrementNumberOfRetries();
+                            return Task.CompletedTask;
+                        });
+                        s.NumberOfRetries(c.NumberOfMessages);
+                    });
+                    recoverability.Delayed(s => s.NumberOfRetries(0));
+                });
+            }
+
+            class HighContentionSaga : Saga<HighContentionSaga.HighContentionSagaData>, IAmStartedByMessages<StartSaga>, IHandleMessages<AdditionalMessage>
+            {
+                public Context TestContext { get; set; }
+
+                protected override void ConfigureHowToFindSaga(SagaPropertyMapper<HighContentionSagaData> mapper)
+                {
+                    mapper.ConfigureMapping<StartSaga>(m => m.SomeId).ToSaga(d => d.SomeId);
+                    mapper.ConfigureMapping<AdditionalMessage>(m => m.SomeId).ToSaga(d => d.SomeId);
+                }
+
+                public Task Handle(StartSaga message, IMessageHandlerContext context)
+                {
+                    Data.SomeId = message.SomeId;
+                    TestContext.Watch.Start();
+                    TestContext.SagaStarted = true;
+
+                    return context.SendLocal(new FireInTheWhole { SomeId =  message.SomeId });
+                }
+
+                public class HighContentionSagaData : ContainSagaData
+                {
+                    public int Hit { get; set; }
+                    public Guid SomeId { get; set; }
+                }
+
+                public async Task Handle(AdditionalMessage message, IMessageHandlerContext context)
+                {
+                    Data.Hit++;
+
+                    if (Data.Hit >= TestContext.NumberOfMessages)
+                    {
+                        MarkAsComplete();
+                        await context.SendLocal(new DoneSaga { SomeId = message.SomeId, HitCount = Data.Hit });
+                    }
+                }
+            }
+
+            class CreateLoadHandler : IHandleMessages<FireInTheWhole>
```

Comment:
not sure we need this. The saga should always be saved before we dispatch the messages so it should be fine to do this in the saga-started handler. Otherwise, I'd suggest adding a comment to document the reason for this handler.

Reply:
Done